capellambse/extensions/validation/_validate.py

```diff
@@ -0,0 +1,483 @@
+            pass
+
+        _, class_ = c.resolve_handler(key)
+        return RealType(class_)
+
+    def register(self, vtype: VirtualType) -> None:
+        if vtype.name in self:
+            known = self.__registry[vtype.name]
+            raise RuntimeError(f"Virtual type already known: {known}")
+        self.__registry[vtype.name] = vtype
+
+
+_types_registry = _VirtualTypesRegistry()
+
+
+def virtual_type(
+    real_type: str | type[_T],
+) -> cabc.Callable[[cabc.Callable[[_T], bool]], VirtualType[_T]]:
+    def decorate(func: cabc.Callable[[_T], bool]) -> VirtualType[_T]:
+        nonlocal real_type
+        if isinstance(real_type, str):
+            _, real_type = c.resolve_handler(real_type)
+
+        vtype = VirtualType(func.__name__, real_type, func)
+        _types_registry.register(vtype)
+        return vtype
+
+    return decorate
+
+
+@virtual_type(model.oa.OperationalActivity)
+def OperationalActivity(obj):
+    return obj.parent != obj._model.oa.activity_package
+
+
+@virtual_type(model.ctx.SystemFunction)
+def SystemFunction(obj):
+    return obj.parent != obj._model.sa.function_package
+
+
+@virtual_type(model.la.LogicalFunction)
+def LogicalFunction(obj):
+    return obj.parent != obj._model.la.function_package
+
+
+@virtual_type(model.la.LogicalFunction)
+def PhysicalFunction(obj):
+    return obj.parent != obj._model.pa.function_package
+
+
+class Category(mt._StringyEnumMixin, enum.Enum):
+    """A category for a rule."""
+
+    REQUIRED = enum.auto()
+    RECOMMENDED = enum.auto()
+    SUGGESTED = enum.auto()
+
+
+@dataclasses.dataclass
+class Result:
+    """A validation rule result."""
+
+    uuid: helpers.UUIDString
+    category: Category
+    value: bool
+    object: c.GenericElement
+
+    def __repr__(self) -> str:
+        """Return the representation of a result."""
+        obj_repr = self.object._short_repr_()
+        return (
+            f"Result(uuid={self.uuid!r}, category={self.category!r},"
+            f" value={self.value!r}, object={obj_repr})"
+        )
+
+
+ElementType = t.TypeVar("ElementType", bound=c.GenericElement)
+
+
+@dataclasses.dataclass(frozen=True)
+class Rule(t.Generic[_T]):
+    """Common (generic) class representing any validation rule."""
+
+    id: str
+    name: str
+    types: list[str]
+    rationale: str
+    category: Category
+    action: str
+    validator: cabc.Callable[[_T], bool]
+    """Returns True if the object passed this rule."""
+    hyperlink_further_reading: str | None = None
+
+    def find(self, model_: capellambse.MelodyModel) -> cabc.Iterator[_T]:
+        for i in self.types:
+            yield from _types_registry[i].search(model_)
+
+    def __call__(self, obj: _T) -> bool:
+        return self.validator(obj)
+
+    def __hash__(self) -> int:
+        return hash(self.id)
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, str):
+            return self.id == other
+        return hash(self) == hash(other)
```

Comment:
This is just plain wrong. `hash(a) == hash(b)` **does not** imply that `a == b`, only the other way around.